### DISCUSSION TITLE
An abstraction to retrieve task results from task IDs

### DISCUSSION THREAD

**Discussion by ksamuel** [RealOrangeOne/django-tasks#62] [Ideas]
Reading the DEP, I can't see an API clearly defined to retrieve a task using its id.

Indeed:

- We may request the task status from a different process, knowing only the ID.
- We can get the information in a backend specific way (e.g: orm request), but it's not generic, which prevent easy testing or creating pluggable apps that always work no matter the backend.
- We could imagine creating a TaskResult and call refresh(), but the parameters other than ID seems mandatory in the DEP and we don't know them yet. Also, TaskResult is a base class, but we need to call it from an instance of a concrete backend specific TaskResult child, which again is not generic.

E.G:

Let's say I start a background task from an AJAX request, and now I do long polling from the client to know the task status. On a WSGI app, I don't know what process I'll end up with, so I can't cache the task result object on the backend, I can only keep the task ID on the frontend.

I need a way to say something like:

```
from django.task import TaskResult

TaskResult.from_id(id=...)
```

As well as the shortcuts like get_object_or_404, but for a task.

**Comment by RealOrangeOne** [comment]
This is currently possible, but off the backend instead:

```python
from django_tasks import default_task_backend 

default_task_backend.get_result(...)
```

Alternatively, you can do it from the task:

```python
from tasks import my_task

my_task.get_result(...)
```

The latter doing an internal check to make sure the result is for the correct `Task`. Adding it to a `TaskResult` is probably too many interfaces, not to mention adds complexity when using multiple backends.

It's only mentioned in passing in the DEP (see the interfaces for task and backend), but it's absolutely a required feature I agree.

Shortcuts is an interesting idea, but I suspect it's less useful as the ORM counterparts. The result ID is generally only useful internally, so 404-ing when the result is missing is rarely the right way to go.

**Reply by ksamuel** [reply]
I missed the get_result on the backed, my bad. As long as we can do it from the backend, it's all good.

 The task won't work, since you will likely not have a reference to the task instance anymore when the new request to get the status is made.

Thanks for answering and closing.
